src/Avalonia.Base/Threading/ReaderWriterLockWrapper.cs

```diff
@@ -0,0 +1,100 @@
+using System;
+using System.Threading;
+
+namespace Avalonia.Threading;
+
+/// <summary>
+///     A reader-writer lock abstraction, wrapping around the .NET class ReaderWriterLockSlim.
+/// </summary>
+public class ReaderWriterLockWrapper
+{
+    /// <summary>
+    ///     Executes an action within a read lock.
+    /// </summary>
+    /// <param name="action">The action to execute</param>
+    public void ExecuteWithinReadLock(Action action)
```

Comment:
I can't review this code confidently, but ReaderWriterLockWrapper abstraction will introduce unnecessary  closures and allocations in the hot path. It also should be internal.

Reply:
Thank you for the feedback, @maxkatz6!

I have updated the PR accordingly.

Reply:
The closures are still here though: as mentioned by Max, we should avoid them.

In general, there should be almost no contention here, I don't think the RWL buys us anything compared to a simple lock in this situation.

Reply:
Thank you for the update, @MrJul!

When changing back to the lock() mechanism, should I also wrap the getter of the property **IsEnabled** within the lock object, or leave it as before, with only the volatile modifier of the backing field ensuring consistency?

Reply:
to avoid the closure you can do change like this or add overload
```suggestion
    public void ExecuteWithinReadLock<T>(Action<T> action)
```

and use like:

```c#
    public bool IsEnabled
    {
        get => _readerWriterLockWrapper.ExecuteWithinReadLock(() => _isEnabled);
        set
        {
            _readerWriterLockWrapper.ExecuteWithinWriteLock(state =>
            {
                if (!state && _isEnabled)
                {
                    Start();
                }
            }, value);
        }
    }
```

Reply:
Thank you for the suggestion, @workgroupengineering!

As far as I understand, the closures will remain, as the compiler creates, behind the scenes, a class for every anonymous method, even if the method does not reference its outer context.

Anyway, in this case, the only advantage of the RWL over lock() is allowing multiple simultaneous readers. If this is not a likely scenario, then the mutual exclusion of lock() would be sufficient.

Reply:
@mihnea-radulescu
If a lambda doesn't reference a variable, it will be allocated only once (during a program lifetime), however when it does reference outside variables, an allocation occurs each time such a lambda is created - as it needs to capture the current state. Hence the Action<T> trick, so that the lambda doesn't need to reference outside variables - it can just access the parameter.

Reply:
Thank you for the explanation, @BAndysc!